eng/Tools.props

```diff
@@ -1,14 +1,25 @@
 <Project>
-  <ItemGroup>
+  <ItemGroup Condition=" '$(DotNetBuildFromSource)' != 'true' ">
     <!--
-      This is here to workaround flakiness in the NuGet SDK resolver in MSBuild.
-      Arcade will run a pre-restore for these packages. This works more consistently than the SDK resolution which uses global.json.
-      Without this here, we see regular failures with 'error MSB4236: The SDK 'Yarn.MSBuild' specified could not be found.'
-      Since this project is evaluated before .npmproj files are loaded, this should cause the package to end
-      up in the NuGet cache ahead of time.
-
-      This is not needed in source build.
+      This is here to workaround flakiness in the NuGet SDK resolver in MSBuild. Arcade will run a pre-restore for
+      these packages. This works more consistently than the SDK resolution which uses global.json. Without this
+      here, we see regular failures with 'error MSB4236: The SDK 'Yarn.MSBuild' specified could not be found.'
+      Since this project is evaluated before .npmproj files are loaded, this should cause the package to end up in
+      the NuGet cache ahead of time. This is not needed in source build.
     -->
-    <PackageReference Condition="'$(DotNetBuildFromSource)' != 'true'" Include="Yarn.MSBuild" Version="1.15.2" />
+    <PackageReference Include="Yarn.MSBuild" Version="1.15.2" />
   </ItemGroup>
+
+  <!-- Update the generated files when we restore projects. Skip in desktop msbuild due to VS 16.8 requirements. -->
+  <Target Name="GenerateDirectoryBuildFiles"
+      AfterTargets="Restore"
+      Condition=" '$(DotNetBuildFromSource)' != 'true' AND '$(MSBuildRuntimeType)' == 'core' ">
+    <!-- Separate invocations and use different properties to ensure second can load the restored package info. -->
+    <MSBuild Projects="$(RepoRoot)eng\tools\GenerateFiles\GenerateFiles.csproj"
+        RemoveProperties="BaseIntermediateOutputPath"
+        Targets="Restore" />
+    <MSBuild Projects="$(RepoRoot)eng\tools\GenerateFiles\GenerateFiles.csproj"
+        RemoveProperties="BaseIntermediateOutputPath;ExcludeRestorePackageImports"
+        Targets="GenerateDirectoryBuildFiles" />
+  </Target>
```

Comment:
@dotnet/aspnet-build if you have a chance, please dummy-check my many manual changes in this PR. Big example here and in the new eng/tools/GenerateFiles/ project.